src/asmcnc/apps/drywall_cutter_app/widget_drywall_shape_display.py

```diff
@@ -298,14 +411,209 @@ def rotation_required(self):
+        d_limit = min(x_machine_range, y_machine_range)
+        if current_shape == 'circle' and float(self.d_input.text or 0) > d_limit:
+            self.d_input_validation_label.text = 'MAX: ' + str(d_limit)
+            self.d_input_validation_label.opacity = 1
+        else:
+            self.d_input_validation_label.opacity = 0
+
+        if current_shape in ['square', 'rectangle']:
+            x_limit = x_practical_range
+            y_limit = y_practical_range
+            r_limit = min(x_limit, y_limit) / 2
+            # Because square is limited by the smaller dimension
+            square_limit = min(x_limit, y_limit)
+
+            if current_shape == 'square':
+                self.x_input_validation_label.opacity = 0
+                if float(self.y_input.text or 0) > square_limit:
+                    self.y_input_validation_label.text = 'MAX: ' + str(square_limit)
+                    self.y_input_validation_label.opacity = 1
+                else:
+                    self.y_input_validation_label.opacity = 0
+            else:
+                if float(self.x_input.text or 0) > x_limit:
+                    self.x_input_validation_label.text = 'MAX: ' + str(x_limit)
+                    self.x_input_validation_label.opacity = 1
+                else:
+                    self.x_input_validation_label.opacity = 0
+
+                if float(self.y_input.text or 0) > y_limit:
+                    self.y_input_validation_label.text = 'MAX: ' + str(y_limit)
+                    self.y_input_validation_label.opacity = 1
+                else:
+                    self.y_input_validation_label.opacity = 0
+
+            if float(self.r_input.text or 0) > r_limit:
+                self.r_input_validation_label.text = 'MAX: ' + str(r_limit)
+                self.r_input_validation_label.opacity = 1
+            else:
+                self.r_input_validation_label.opacity = 0
+        else:
+            self.r_input_validation_label.opacity = 0
+            self.x_input_validation_label.opacity = 0
+            self.y_input_validation_label.opacity = 0
+
+        if current_shape == 'line':
+            if "horizontal" in self.shape_dims_image.source:
+                if float(self.l_input.text or 0) > y_practical_range:
+                    self.l_input_validation_label.text = 'MAX: ' + str(y_practical_range)
+                    self.l_input_validation_label.opacity = 1
+                else:
+                    self.l_input_validation_label.opacity = 0
+            else:
+                if float(self.l_input.text or 0) > x_practical_range:
+                    self.l_input_validation_label.text = 'MAX: ' + str(x_practical_range)
+                    self.l_input_validation_label.opacity = 1
+                else:
+                    self.l_input_validation_label.opacity = 0
+        else:
+            self.l_input_validation_label.opacity = 0
+
+    def are_inputs_valid(self):
```

Comment:
same as above, could this be broken out into separate functions to check validity of components?